File: sdk/ruby/lib/experimentation_platform/http_client.rb
require 'net/http'
require 'uri'
require 'json'

module ExperimentationPlatform
  # Lightweight HTTP client wrapping Net::HTTP.
  #
  # All requests include an Authorization header derived from the configured API key.
  # Responses are parsed as JSON. Non-2xx responses raise APIError subclasses.
  # Network failures raise NetworkError.
  class HttpClient
    # @param base_url [String] API base URL (e.g. "https://api.example.com")
    # @param api_key  [String] API key sent as "Bearer <api_key>"
    # @param timeout  [Integer] open/read timeout in seconds (default: 10)
    def initialize(base_url:, api_key:, timeout: 10)
      @base_url = base_url.chomp('/')
      @api_key  = api_key
      @timeout  = timeout.to_i
    end

    # Perform a GET request.
    #
    # @param path    [String] URL path (e.g. "/api/v1/sdk/flags/my-flag")
    # @param headers [Hash]   additional HTTP headers
    # @return [Hash, Array] parsed JSON response body
    # @raise [AuthenticationError] on 401
    # @raise [APIError]            on other 4xx/5xx
    # @raise [NetworkError]        on connection/timeout failures
    def get(path, headers: {})
      request(:get, path, nil, headers)
    end

    # Perform a POST request with a JSON body.
    #
    # @param path    [String] URL path
    # @param body    [Hash]   request body (serialized to JSON)
    # @param headers [Hash]   additional HTTP headers
    # @return [Hash, Array] parsed JSON response body
    # @raise [AuthenticationError] on 401
    # @raise [APIError]            on other 4xx/5xx
    # @raise [NetworkError]        on connection/timeout failures
    def post(path, body, headers: {})
      request(:post, path, body, headers)
    end

    private

    def request(method, path, body, extra_headers)
      uri = URI.parse("#{@base_url}#{path}")

      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl     = (uri.scheme == 'https')
      http.open_timeout = @timeout
      http.read_timeout = @timeout

      req_class = method == :get ? Net::HTTP::Get : Net::HTTP::Post
      req = req_class.new(uri.request_uri)

      # Default headers
      req['Authorization'] = "Bearer #{@api_key}"
      req['Accept']        = 'application/json'
      req['Content-Type']  = 'application/json'
      req['User-Agent']    = "ExperimentationPlatform-Ruby/#{ExperimentationPlatform::VERSION}"

      extra_headers.each { |k, v| req[k] = v }

      if body
        req.body = body.is_a?(String) ? body : JSON.generate(body)
      end

      response = http.request(req)
      handle_response(response)
    rescue Net::OpenTimeout, Net::ReadTimeout, Errno::ETIMEDOUT => e
      raise NetworkError, "Request timed out: #{e.message}"
    rescue Errno::ECONNREFUSED, Errno::EHOSTUNREACH, Errno::ECONNRESET,
           Errno::ENETUNREACH, SocketError => e
      raise NetworkError, "Connection failed: #{e.message}"
    rescue NetworkError, APIError, AuthenticationError
      raise
    rescue StandardError => e
      raise NetworkError, "Unexpected network error: #{e.message}"
    end

    def handle_response(response)
      code = response.code.to_i

      case code
      when 200..299
        body = response.body
        return {} if body.nil? || body.strip.empty?

        begin
          JSON.parse(body, symbolize_names: false)
        rescue JSON::ParserError
          {}
        end
      when 401
        msg = extract_error_message(response) || "Authentication failed"
        raise AuthenticationError.new(msg, status_code: 401)
      else
        msg = extract_error_message(response) || "API error (HTTP #{code})"
        raise APIError.new(msg, status_code: code)
      end
    end

    def extract_error_message(response)
      body = response.body
      return nil if body.nil? || body.strip.empty?

      parsed = JSON.parse(body)
      parsed['detail'] || parsed['message'] || parsed['error'] || body
    rescue JSON::ParserError
      body
    end
  end
end

File: sdk/elixir/lib/experimentation_platform/http_client.ex
defmodule ExperimentationPlatform.HttpBehaviour do
  @moduledoc """
  Behaviour defining the HTTP client interface.

  Implement this to provide a custom or mock HTTP client (pass the module as
  the `:http_client` config option).
  """

  @callback get(config :: ExperimentationPlatform.Config.t(), path :: String.t()) ::
              {:ok, map()} | {:error, term()}

  @callback post(
              config :: ExperimentationPlatform.Config.t(),
              path :: String.t(),
              body :: map()
            ) :: {:ok, map()} | {:error, term()}
end

defmodule ExperimentationPlatform.HttpClient do
  @moduledoc """
  HTTP client using OTP's built-in `:httpc` (from `:inets`) for zero external
  dependencies beyond Jason.

  Every request carries the configured API key in the `X-API-Key` header plus
  `Content-Type`/`Accept: application/json`. Responses are decoded as JSON.

  ## Error Tuples

  - `{:error, {:auth_error, 401}}` — 401 Unauthorized (bad API key)
  - `{:error, {:api_error, status_code, body}}` — other 4xx/5xx
    (404 experiment/flag not ACTIVE, 422 validation, 429 rate limited)
  - `{:error, {:network_error, reason}}` — connection failures and timeouts
  """

  @behaviour ExperimentationPlatform.HttpBehaviour

  alias ExperimentationPlatform.Config

  @user_agent ~c"ExperimentationPlatform-Elixir-SDK/0.1.0"

  @doc """
  Perform a GET request to the given path (which may include a query string).

  ## Returns
  - `{:ok, map()}` on success (2xx)
  - `{:error, {:auth_error, 401}}` on authentication failure
  - `{:error, {:api_error, status_code, body}}` on other HTTP errors
  - `{:error, {:network_error, reason}}` on network failure
  """
  @impl true
  @spec get(Config.t(), String.t()) :: {:ok, map()} | {:error, term()}
  def get(%Config{} = config, path) do
    ensure_started()
    request = {build_url(config.base_url, path), build_headers(config.api_key)}

    :get
    |> :httpc.request(request, http_options(config.timeout), request_options())
    |> handle_result()
  end

  @doc """
  Perform a POST request to the given path with a JSON body.

  ## Returns
  - `{:ok, map()}` on success (2xx)
  - `{:error, {:auth_error, 401}}` on authentication failure
  - `{:error, {:api_error, status_code, body}}` on other HTTP errors
  - `{:error, {:network_error, reason}}` on network failure
  """
  @impl true
  @spec post(Config.t(), String.t(), map()) :: {:ok, map()} | {:error, term()}
  def post(%Config{} = config, path, body) do
    ensure_started()

    encoded_body =
      case Jason.encode(body) do
        {:ok, json} -> json
        {:error, reason} -> raise ArgumentError, "Failed to encode request body: #{inspect(reason)}"
      end

    request =
      {build_url(config.base_url, path), build_headers(config.api_key), ~c"application/json",
       encoded_body}

    :post
    |> :httpc.request(request, http_options(config.timeout), request_options())
    |> handle_result()
  end

  @doc false
  @spec build_url(String.t(), String.t()) :: charlist()
  def build_url(base_url, path) do
    base = String.trim_trailing(base_url, "/")
    normalized_path = if String.starts_with?(path, "/"), do: path, else: "/" <> path
    to_charlist("#{base}#{normalized_path}")
  end

  @doc false
  @spec build_headers(String.t()) :: [{charlist(), charlist()}]
  def build_headers(api_key) do
    [
      {~c"X-API-Key", to_charlist(api_key)},
      {~c"Content-Type", ~c"application/json"},
      {~c"Accept", ~c"application/json"},
      {~c"User-Agent", @user_agent}
    ]
  end

  # --- Private Helpers ---

  defp ensure_started do
    {:ok, _} = Application.ensure_all_started(:inets)
    {:ok, _} = Application.ensure_all_started(:ssl)
    :ok
  end

  defp http_options(timeout), do: [timeout: timeout, connect_timeout: timeout]

  defp request_options, do: [body_format: :binary]

  defp handle_result({:ok, {{_version, status, _reason}, _resp_headers, body}}) do
    handle_response(status, body)
  end

  defp handle_result({:error, reason}), do: {:error, {:network_error, reason}}

  defp handle_response(status, body) when status >= 200 and status < 300 do
    body_str = to_string(body)

    case Jason.decode(body_str) do
      {:ok, decoded} when is_map(decoded) -> {:ok, decoded}
      {:ok, decoded} -> {:ok, %{"data" => decoded}}
      {:error, _} -> {:ok, %{"raw" => body_str}}
    end
  end

  defp handle_response(401, _body), do: {:error, {:auth_error, 401}}

  defp handle_response(status, body), do: {:error, {:api_error, status, to_string(body)}}
end
